File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        
        int[] ans = new int[nums.length];
        
        int zerocount = 0;
        int product = 1;

        for(int i=0;i<nums.length;i++)
        {
            if(nums[i]==0)
            {
                zerocount++;
            }
            else
            {
                product = product*nums[i];
            }
        }

        if(zerocount == 0)
        {
            for(int j=0;j<nums.length;j++)
            {
                ans[j] = product/nums[j];
            }

        }
        else if(zerocount == 1)
        {
            for(int j=0;j<nums.length;j++)
            {
                if(nums[j]==0)
                {
                    ans[j] = product;
                }
                
            }
        }

        return ans;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        prefix = [1] * len(nums)

        for i in range(1,len(nums)):
            prefix[i] = prefix[i-1] * nums[i-1]
        
        postfix = [1] * len(nums)
        
        for i in range(len(nums)-2,-1,-1):
            postfix[i] = postfix[i+1] * nums[i+1]

        res = []
        for i in range(0,len(nums)):
            res.append(prefix[i] * postfix[i])

        return res




       

        